contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -113,6 +117,86 @@ convert_blob(struct agent_connection* con, const char *blob, DWORD blen, char **
 	return success? 0: -1;
 }
 
+/*
+ * in current_user sub tree under key_name key
+ * remove all sub keys with value name value_name_to_remove
+ * and value data value_data_to_remove
+ */
+static int
+remove_matching_subkeys_from_registry(HKEY user_root, wchar_t* key_name, wchar_t* value_name_to_remove, char *value_data_to_remove) {
+	int index = 0, success = 0;
+	DWORD data_len;
+	HKEY root = 0, sub = 0;
+	char *data = NULL;
+	wchar_t sub_name[MAX_KEY_LENGTH];
+	DWORD sub_name_len = MAX_KEY_LENGTH;
+
+	if (RegOpenKeyExW(user_root, key_name, 0, DELETE | KEY_ENUMERATE_SUB_KEYS | KEY_WOW64_64KEY, &root) != 0) {
+		success = 1;
+		goto done;
+	}
+
+	while (1) {
+		sub_name_len = MAX_KEY_LENGTH;
+		if (sub) {
+			RegCloseKey(sub);
+			sub = NULL;
+		}
+		if (RegEnumKeyExW(root, index++, sub_name, &sub_name_len, NULL, NULL, NULL, NULL) == 0) {
+			if (RegOpenKeyExW(root, sub_name, 0, KEY_QUERY_VALUE | KEY_WOW64_64KEY, &sub) == 0 &&
+				RegQueryValueExW(sub, value_name_to_remove, 0, NULL, NULL, &data_len) == 0) {
+
+				if (data)
+					free(data);
+				data = NULL;
+
+				if ((data = malloc(data_len + 1)) == NULL ||
```

Comment:
You are correct that in our specific use case of this function, `data_len` is a user-controlled value (the canonical path to the provider and the pin).
Since this is a utility function, I think it would make sense for it not to assume any specific use case, try to allocate the desired memory, and fail if the allocation fails without leaking memory.
Bound checking is still a good idea, though. I suggest we do that in an upstream function, when values are being written (`process_add_smartcard_key()`)
Bound checking on the canonical path is already being made by `realpath()` which would ensure that it will not exceed `PATH_MAX`. I've added a bound check on the pin length.
Please let me know if that makes sense.

Reply:
I'd still recommend adding a check in `remove_matching_subkeys_from_registry()` to make the code more resilient toward unintended changes in `process_add_smartcardkey()` but hold no objections.

Reply:
Sure. I've added a bound check per Microsoft's [recommendation](https://docs.microsoft.com/en-us/windows/win32/sysinfo/registry-element-size-limits) not to use data values longer than 2KB.